### PR TITLE
AML-48 Set up API

### PR DESCRIPTION
**Pull request by tstreule**
Set ground stones for Amplo's API.

**Goal:** Automatic synchronization with Azure blob storage, model training and uploading to Amplo's dashboard.

Additionally, fixed a bug in `IntervalAnalyzer` and added kwarg for `Pipeline.__init__` (model_dir)

### CODE REVIEW

File: Amplo/Pipeline.py [hunk 1]
```diff
@@ -472,8 +473,10 @@ def _read_data(self, *args, **kwargs) -> pd.DataFrame:
                 # Do nothing
                 pass
             elif os.path.isdir(data):
+                # Interval-analyze data
                 data = IntervalAnalyser(folder=data).fit_transform()
-                self.target = 'labels'
+                # Set target it not already set manually
+                self.target = self.target or 'labels'
```
Comment (nielsuit227) [inline Amplo/Pipeline.py:479]: This should always be the case? If there Interval Analyser is used while a target is set, it still comes out as labels?

File: Amplo/Pipeline.py [hunk 2]
```diff
@@ -472,8 +473,10 @@ def _read_data(self, *args, **kwargs) -> pd.DataFrame:
                 # Do nothing
                 pass
             elif os.path.isdir(data):
+                # Interval-analyze data
```
Comment (nielsuit227) [inline Amplo/Pipeline.py:476]: Can you store this? For version control, and then check if it's there to avoid duplicate runs of the interval analyser?

File: AmploAPI/interface.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+from termcolor import cprint
+
+import pandas as pd
+from Amplo import Pipeline
+from Amplo.AutoML import IntervalAnalyser
+
+from sync import synchronize_data
+
+
+__all__ = ['AmploAPI']
```
Comment (nielsuit227) [inline AmploAPI/interface.py:13]: What's does this do?

File: AmploAPI/interface.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+
+class AmploAPI:
```
Comment (nielsuit227) [inline AmploAPI/interface.py:16]: Let's rename that to `API`, then you can import it as `from Amplo.API import x`.

File: AmploAPI/sync.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    """
+
+    # Params
+    team = team or input('Please enter the TEAM: ')
+    service = service or input('Please enter the SERVICE: ')
+    client = BlobServiceClient.from_connection_string(os.getenv('AZURE_STORAGE_STRING'))
+    container = client.get_container_client('amploplatform')
+
+    # Get machines
+    machines = get_folders(team, container)
```
Comment (nielsuit227) [inline AmploAPI/sync.py:64]: Pretty straight forward to support a provided list of machines instead as well, let's do that directly.

File: AmploAPI/interface.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        path_to_latest = max(all_versions)
+        data = pd.read_csv(path_to_latest, index_col=[0, 1])
+
+        # Drop unnamed columns if any
+        unnamed_cols = []
+        for col_name in data.columns:
+            if 'unnamed' in col_name.lower():
+                unnamed_cols += col_name
+        if len(unnamed_cols) > 0:
+            warnings.warn(f'[AmploML] Unnamed columns were found and removed, namely: {unnamed_cols}')
```
Comment (nielsuit227) [inline AmploAPI/interface.py:260]: It's actually `[AutoML]`, but love this

File: AmploAPI/interface.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+        all_versions = list(Path(data_dir).glob('*.csv'))
+        if len(all_versions) == 0:
+            # There is nothing to read
+            return None
+
+        # Read data
+        path_to_latest = max(all_versions)
+        data = pd.read_csv(path_to_latest, index_col=[0, 1])
+
+        # Drop unnamed columns if any
```
Comment (nielsuit227) [inline AmploAPI/interface.py:254]: This doesn't belong here. Removing unnamed columns is great, but should happen in the `DataProcesser`. The `IntervalAnalyser` uses the `DataProcesser`, and we load data from that, so there is no point of failure other than our code.

File: AmploAPI/interface.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+            warnings.warn(f'[AmploML] Unnamed columns were found and removed, namely: {unnamed_cols}')
+            data.drop(unnamed_cols, axis=1, inplace=True)
+
+        return data
+
+
+if __name__ == '__main__':
+    # Example of use
+    sync_dir = Path('../Sandbox')  # mother directory for synchronization
+    data_select = dict(teams='Tritium', machines=None, services='Diagnostics')
```
Comment (nielsuit227) [inline AmploAPI/interface.py:269]: What is this? This shouldn't even need a main? This should be separately tested in `tests`.

**Review by nielsuit227 — COMMENTED** [review]
The `Pipeline` is continued as usual, yet in the `IntervalAnalyser`, the data is already cleaned and features extracted. Would it be possible to avoid rerunning them in the pipeline after the `IntervalAnalyser`?

File: AmploAPI/interface.py (hunk 3 above)
Comment (tstreule) [reply AmploAPI/interface.py:13]: It is a list of strings defining what symbols in a module will be exported when `from <module> import *` is used on the module.

[See on Stackoverflow](https://stackoverflow.com/questions/44834/can-someone-explain-all-in-python)

Comment (nielsuit227) [reply AmploAPI/interface.py:13]: Cool stuff, didn't know that!

File: AmploAPI/interface.py (hunk 8 above)
Comment (tstreule) [reply AmploAPI/interface.py:269]: Removed and created test cases
cf. commit abbdda0d0a1513e0eebc0f376ea514cb202c5fb1

File: AmploAPI/interface.py (hunk 7 above)
Comment (tstreule) [reply AmploAPI/interface.py:254]: Removed in commit abbdda0d0a1513e0eebc0f376ea514cb202c5fb1

File: AmploAPI/interface.py (hunk 6 above)
Comment (tstreule) [reply AmploAPI/interface.py:260]: This part was anyway removed in commit abbdda0d0a1513e0eebc0f376ea514cb202c5fb1

File: AmploAPI/interface.py (hunk 4 above)
Comment (tstreule) [reply AmploAPI/interface.py:16]: Good point! In commit abbdda0d0a1513e0eebc0f376ea514cb202c5fb1 it was renamed accordingly

File: Amplo/Pipeline.py (hunk 1 above)
Comment (tstreule) [reply Amplo/Pipeline.py:479]: `IntervalAnalyser.target`

File: Amplo/Pipeline.py (hunk 2 above)
Comment (tstreule) [reply Amplo/Pipeline.py:476]: `try except`

Comment (tstreule) [reply Amplo/Pipeline.py:476]: See commit ca976d06ae752899e5c2571fbba5503f7c2ff2a4

File: Amplo/Pipeline.py (hunk 1 above)
Comment (tstreule) [reply Amplo/Pipeline.py:479]: See commit ca976d06ae752899e5c2571fbba5503f7c2ff2a4

**Comment by tstreule** [comment]
> The `Pipeline` is continued as usual, yet in the `IntervalAnalyser`, the data is already cleaned and features extracted. Would it be possible to avoid rerunning them in the pipeline after the `IntervalAnalyser`?

See commit 9cc5430c8d65f78837185eab20a0cdd294b3b2c6
